snuba/datasets/__init__.py

```diff
@@ -33,12 +39,55 @@ def get_writer(self, options=None, table_name=None):
         from snuba import settings
         from snuba.clickhouse.http import HTTPBatchWriter
 
+        def default(value):
+            if isinstance(value, datetime):
+                return value.strftime(DATETIME_FORMAT)
+            else:
+                raise TypeError
+
+        def encode(rows: Iterable[WriterTableRow]) -> Iterable[bytes]:
+            return map(
+                lambda row: json.dumps(row, default=default).encode("utf-8"),
+                rows,
+            )
+
+        return HTTPBatchWriter(
+            self._schema,
+            settings.CLICKHOUSE_HOST,
+            settings.CLICKHOUSE_HTTP_PORT,
+            encode,
+            options,
+            table_name,
+        )
+
+    def get_bulk_writer(self, options=None, table_name=None):
+        """
+        This is a stripped down verison of the writer designed
+        for better performance when loading data in bulk.
+        """
+        # TODO: Consider using rapidjson to encode everywhere
+        # once we will be confident it is reliable enough.
+
+        from snuba import settings
+        from snuba.clickhouse.http import HTTPBatchWriter
+
+        def encode(rows: Iterable[WriterTableRow]) -> Iterable[bytes]:
+            ret = bytearray()
+            for row in rows:
+                ret += rapidjson.dumps(row).encode("utf-8")
```

Comment:
From my understanding, this isn't an efficient way of building strings (at least it wasn't in Python 2.)

Did you benchmark this to come to this implementation?

Reply:
Doing this slashed the time from 90 minutes to 50.  But that is due to the fact that we provide one big byte array at every call instead of an iterable where every entry is a row. 
The overhead of the multiple chunks was much bigger than the overhead of building the string as a bytearray. I did not look into how much overhead the bytearray is responsible for after the change.

Reply:
So this is the impact:
now encode takes 0.444 cumulative seconds over a total 4 seconds.
in encode most of the time is taken by rapidjson with 0.338 of cumulative time, so improving the string building can help with 0.1 seconds.

Reply:
Tried with .join
```
ret = "".join(rapidjson.dumps(r) for r in rows)
            return [ret.encode("utf-8")]
```

Which is suppsoed to be the right way in python3.5 and above but I got the exact same result from the benchmark